File: pot_scraper/modernize.py
"""On-demand AI modernization of a single old recipe.

Free scoring covers the whole library; this step only runs when you pick a
recipe you actually want to cook. Uses Claude to convert archaic measures and
wood-stove directions into modern kitchen steps plus a Charleston shopping
list. Requires the `anthropic` package and ANTHROPIC_API_KEY.
"""

from typing import List
from .config import DEFAULT_MODEL
from .pantry import WALMART, ALDI, COSTCO


class ModernizeError(RuntimeError):
    pass


def _build_schema():
    """Import pydantic lazily so the core CLI works without it."""
    from pydantic import BaseModel, Field

    class ShopItem(BaseModel):
        item: str = Field(description="Grocery item with quantity, e.g. '2 lb beef chuck'")
        store: str = Field(description=f"One of: {WALMART}, {ALDI}, {COSTCO}")

    class ModernRecipe(BaseModel):
        title: str
        servings: str = Field(description="Modern serving estimate, e.g. '4-6'")
        ingredients: List[str] = Field(description="Modern ingredient list with US measures")
        steps: List[str] = Field(description="Numbered-in-order modern cooking steps, oven temps in Fahrenheit")
        shopping_list: List[ShopItem] = Field(description="What to buy, mapped to a Charleston store")
        notes: str = Field(description="Substitutions for anything archaic or hard to source")

    return ModernRecipe


_PROMPT = """You are modernizing a historical recipe for a home cook in Charleston, SC \
who shops at Walmart, Aldi, and Costco.

Convert the recipe below into a modern version:
- Translate archaic measures (gill, teacupful, "butter the size of an egg", peck) to US cups/tbsp/tsp/lb.
- Convert wood-stove directions ("slow oven", "quick oven", "hob") to modern oven temps in Fahrenheit and stovetop instructions.
- Replace or substitute hard-to-source ingredients (suet, isinglass, saleratus, etc.) with items available at Walmart/Aldi/Costco, and explain the swap in notes.
- Keep the dish faithful to the original; do not invent a different recipe.
- Build a shopping list mapping each ingredient to ONE store (Walmart, Aldi, or Costco).

ORIGINAL TITLE: {title}
SOURCE: {source}

ORIGINAL TEXT:
{body}
"""


def modernize(recipe):
    """Return a dict of the modernized recipe (also cached onto recipe['modernized'])."""
    try:
        import anthropic  # noqa: F401
    except ImportError:
        raise ModernizeError(
            "The 'anthropic' package is required for `modernize`. "
            "Install it with: pip install 'pot-scraper[ai]'"
        )

    import anthropic

    schema = _build_schema()
    client = anthropic.Anthropic()  # reads ANTHROPIC_API_KEY from env

    prompt = _PROMPT.format(
        title=recipe["title"],
        source=f"{recipe.get('source_title', '?')} ({recipe.get('source_author', '?')})",
        body=recipe["body"][:6000],
    )

    try:
        resp = client.messages.parse(
            model=DEFAULT_MODEL,
            max_tokens=4000,
            messages=[{"role": "user", "content": prompt}],
            output_format=schema,
        )
    except anthropic.AuthenticationError:
        raise ModernizeError(
            "ANTHROPIC_API_KEY is missing or invalid. Put it in "
            "~/.pot-scraper/.env  (line:  ANTHROPIC_API_KEY=sk-ant-...) "
            "or export it in your shell."
        )
    except anthropic.APIError as e:
        raise ModernizeError(f"Claude API error: {e}")

    parsed = resp.parsed_output
    if parsed is None:
        raise ModernizeError("Model did not return a parseable recipe (try again).")

    result = parsed.model_dump()
    recipe["modernized"] = result
    return result

File: netlify/functions/modernize.js
// pot*scraper — modernize an old recipe, server-side.
//
// Takes a historical recipe {title, body, source} and asks Claude to convert
// it into modern US measures, Fahrenheit oven temps, substitutions for
// hard-to-source ingredients, and a Charleston (Walmart/Aldi/Costco) shopping
// list. The API key stays here in Netlify env vars — never in the browser.
//
// Set ANTHROPIC_API_KEY in Netlify env vars (Site config -> Environment).

const API = 'https://api.anthropic.com/v1/messages';
const MODEL = 'claude-opus-4-8'; // matches the CLI default; override via env MODEL
const VERSION = '2023-06-01';

const CORS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'Content-Type',
  'Access-Control-Allow-Methods': 'POST, OPTIONS',
  'Content-Type': 'application/json',
};

const SYSTEM = `You are modernizing a historical recipe for a home cook in Charleston, SC who shops at Walmart, Aldi, and Costco.

Convert the recipe faithfully — same dish, not a new one:
- Translate archaic measures (gill, teacupful, "butter the size of an egg", peck, dram) to US cups/tbsp/tsp/lb.
- Convert wood-stove directions ("slow oven", "quick oven", "hob") to modern oven temperatures in Fahrenheit and clear stovetop steps.
- Replace or substitute hard-to-source ingredients (suet, isinglass, saleratus, neat's feet, etc.) with items available at Walmart/Aldi/Costco, and explain each swap in notes.
- Build a shopping list mapping each ingredient to ONE store: Walmart, Aldi, or Costco.
- Keep it practical and unfussy. Do not invent ingredients that aren't implied by the original.`;

const SCHEMA = {
  type: 'object', additionalProperties: false,
  properties: {
    title: { type: 'string' },
    servings: { type: 'string' },
    ingredients: { type: 'array', items: { type: 'string' } },
    steps: { type: 'array', items: { type: 'string' } },
    shopping_list: {
      type: 'array',
      items: {
        type: 'object', additionalProperties: false,
        properties: {
          item: { type: 'string' },
          store: { type: 'string', enum: ['Walmart', 'Aldi', 'Costco'] },
        },
        required: ['item', 'store'],
      },
    },
    notes: { type: 'string' },
  },
  required: ['title', 'servings', 'ingredients', 'steps', 'shopping_list', 'notes'],
};

exports.handler = async (event) => {
  if (event.httpMethod === 'OPTIONS') return { statusCode: 200, headers: CORS, body: '' };
  if (event.httpMethod !== 'POST') return { statusCode: 405, headers: CORS, body: JSON.stringify({ error: 'Method not allowed' }) };

  const key = process.env.ANTHROPIC_API_KEY;
  if (!key) return { statusCode: 500, headers: CORS, body: JSON.stringify({ error: 'ANTHROPIC_API_KEY not set in Netlify env vars' }) };

  let body;
  try { body = JSON.parse(event.body || '{}'); }
  catch { return { statusCode: 400, headers: CORS, body: JSON.stringify({ error: 'Invalid JSON in request body' }) }; }

  const title = String(body.title || '').slice(0, 200);
  const text = String(body.body || '').slice(0, 6000);
  const source = String(body.source || '').slice(0, 200);
  if (!text) return { statusCode: 400, headers: CORS, body: JSON.stringify({ error: 'need a recipe body to modernize' }) };

  const user =
    `ORIGINAL TITLE: ${title}\n` +
    `SOURCE: ${source}\n\n` +
    `ORIGINAL TEXT:\n${text}`;

  const payload = {
    model: process.env.MODEL || MODEL,
    max_tokens: 2500,
    system: SYSTEM,
    messages: [{ role: 'user', content: user }],
    output_config: { format: { type: 'json_schema', schema: SCHEMA } },
  };

  try {
    const r = await fetch(API, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'x-api-key': key, 'anthropic-version': VERSION },
      body: JSON.stringify(payload),
    });
    const raw = await r.text();
    if (!r.ok) return { statusCode: r.status, headers: CORS, body: JSON.stringify({ error: 'Anthropic API error', detail: raw.slice(0, 600) }) };
    const msg = JSON.parse(raw);
    const jsonText = (msg.content || []).filter((b) => b.type === 'text').map((b) => b.text).join('');
    let data;
    try { data = JSON.parse(jsonText); }
    catch { return { statusCode: 502, headers: CORS, body: JSON.stringify({ error: 'model did not return JSON', raw: jsonText.slice(0, 400) }) }; }
    return { statusCode: 200, headers: CORS, body: JSON.stringify({ ok: true, data }) };
  } catch (e) {
    return { statusCode: 500, headers: CORS, body: JSON.stringify({ error: 'Failed to reach Anthropic API: ' + String(e.message || e).slice(0, 300) }) };
  }
};
